File: uat/custom-components/client-mosquitto-c/src/GRPCDiscoveryClient.cpp
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "GRPCDiscoveryClient.h"                /* self class */
#include "logger.h"                             /* logd() */

using grpc::ClientContext;

using ClientControl::Empty;
using ClientControl::RegisterReply;
using ClientControl::RegisterRequest;
using ClientControl::DiscoveryRequest;
using ClientControl::UnregisterRequest;
using ClientControl::OnReceiveMessageRequest;
using ClientControl::MqttConnectionId;
using ClientControl::OnMqttDisconnectRequest;



bool GRPCDiscoveryClient::RegisterAgent(std::string & local_ip) {
    RegisterRequest request;
    request.set_agentid(m_agent_id);

    RegisterReply reply;
    ClientContext context;

    logd("Sending RegisterAgent request with agent_id %s\n", m_agent_id.c_str());
    Status status = m_stub->RegisterAgent(&context, request, &reply);
    bool success = checkStatus(status);
    if (success) {
        local_ip = reply.address();
        return !local_ip.empty();                  // TODO: check IP is valid
    }

    return false;
}

bool GRPCDiscoveryClient::DiscoveryAgent(const char * address, unsigned short port) {
    DiscoveryRequest request;
    request.set_agentid(m_agent_id);
    request.set_address(address);
    request.set_port(port);

    Empty reply;
    ClientContext context;

    logd("Sending DiscoveryAgent request agent_id '%s' host:port %s:%hu\n", m_agent_id.c_str(), address, port);
    Status status = m_stub->DiscoveryAgent(&context, request, &reply);
    return checkStatus(status);
}


bool GRPCDiscoveryClient::UnregisterAgent(const std::string & reason) {
    UnregisterRequest request;
    request.set_agentid(m_agent_id);
    request.set_reason(reason);

    Empty reply;
    ClientContext context;

    logd("Sending UnregisterAgent request agent_id '%s' reason '%s'\n", m_agent_id.c_str(), reason.c_str());
    Status status = m_stub->UnregisterAgent(&context, request, &reply);
    return checkStatus(status);
}

bool GRPCDiscoveryClient::onReceiveMqttMessage(int connection_id, Mqtt5Message * message) {
    OnReceiveMessageRequest request;

    request.set_agentid(m_agent_id);
    MqttConnectionId * mqtt_connection_id = new MqttConnectionId();
    mqtt_connection_id->set_connectionid(connection_id);
    request.set_allocated_connectionid(mqtt_connection_id);

    request.set_allocated_msg(message);

    Empty reply;
    ClientContext context;

    // TODO: log also details of message
    logd("Sending OnReceiveMessage request agent_id '%s' connection_id %d\n", m_agent_id.c_str(), connection_id);
    Status status = m_stub->OnReceiveMessage(&context, request, &reply);
    return checkStatus(status);
}

bool GRPCDiscoveryClient::onMqttDisconnect(int connection_id, Mqtt5Disconnect * disconnect, const char * error) {
    OnMqttDisconnectRequest request;

    request.set_agentid(m_agent_id);
    MqttConnectionId * mqtt_connection_id = new MqttConnectionId();
    mqtt_connection_id->set_connectionid(connection_id);
    request.set_allocated_connectionid(mqtt_connection_id);

    request.set_allocated_disconnect(disconnect);
    if (error) {
        request.set_error(error);
    }

    Empty reply;
    ClientContext context;

    // TODO: also details of disconnect
    logd("Sending OnMqttDisconnect request agent_id '%s' connection_id %d error '%s'\n", m_agent_id.c_str(), connection_id, error);
    Status status = m_stub->OnMqttDisconnect(&context, request, &reply);
    return checkStatus(status);
}

bool GRPCDiscoveryClient::checkStatus(const Status & status) {
    bool ok = status.ok();
    if (!ok) {
        std::string message = status.error_message();
        std::string details = status.error_details();

        loge("gRPC request failed: %d: '%s': '%s'\n", (int)status.error_code(), message.c_str(), details.c_str());
    }

    return ok;
}

File: uat/custom-components/client-java-paho/src/main/java/com/aws/greengrass/testing/mqtt5/client/grpc/GRPCDiscoveryClient.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.aws.greengrass.testing.mqtt5.client.grpc;

import com.aws.greengrass.testing.mqtt.client.DiscoveryRequest;
import com.aws.greengrass.testing.mqtt.client.Mqtt5Disconnect;
import com.aws.greengrass.testing.mqtt.client.Mqtt5Message;
import com.aws.greengrass.testing.mqtt.client.MqttAgentDiscoveryGrpc;
import com.aws.greengrass.testing.mqtt.client.MqttConnectionId;
import com.aws.greengrass.testing.mqtt.client.MqttQoS;
import com.aws.greengrass.testing.mqtt.client.OnMqttDisconnectRequest;
import com.aws.greengrass.testing.mqtt.client.OnReceiveMessageRequest;
import com.aws.greengrass.testing.mqtt.client.RegisterReply;
import com.aws.greengrass.testing.mqtt.client.RegisterRequest;
import com.aws.greengrass.testing.mqtt.client.UnregisterRequest;
import com.aws.greengrass.testing.mqtt5.client.GRPCClient;
import com.aws.greengrass.testing.mqtt5.client.exceptions.GRPCException;
import com.google.protobuf.ByteString;
import io.grpc.Grpc;
import io.grpc.InsecureChannelCredentials;
import io.grpc.ManagedChannel;
import io.grpc.StatusRuntimeException;
import lombok.NonNull;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * Implementation of gRPC client used to discover agent.
 */
class GRPCDiscoveryClient implements GRPCClient {
    private static final Logger logger = LogManager.getLogger(GRPCDiscoveryClient.class);

    private static final String GRPC_REQUEST_FAILED = "gRPC request failed";

    private final String agentId;
    private final ManagedChannel channel;
    private final MqttAgentDiscoveryGrpc.MqttAgentDiscoveryBlockingStub blockingStub;

    /**
     * Create instance of gRPC client for agent discovery.
     *
     * @param agentId id of agent to identify control channel by gRPC server
     * @param address address of gRPC service
     * @throws GRPCException on errors
     */
    GRPCDiscoveryClient(@NonNull String agentId, @NonNull String address) throws GRPCException {
        super();
        this.agentId = agentId;
        this.channel = Grpc.newChannelBuilder(address, InsecureChannelCredentials.create()).build();
        this.blockingStub = MqttAgentDiscoveryGrpc.newBlockingStub(channel);
    }

    /**
     * Register the agent.
     *
     * @return IP address of client as visible by server
     * @throws GRPCException on errors
     */
    String registerAgent() throws GRPCException {
        RegisterRequest request = RegisterRequest.newBuilder()
                                        .setAgentId(agentId)
                                        .build();
        RegisterReply reply;

        try {
            reply = blockingStub.registerAgent(request);
        } catch (StatusRuntimeException ex) {
            logger.atError().withThrowable(ex).log(GRPC_REQUEST_FAILED);
            throw new GRPCException(ex);
        }
        return reply.getAddress();
    }

    /**
     * Discover the agent.
     *
     * @param address host of local gRPC service
     * @param port port of local gRPC service
     */
    void discoveryAgent(String address, int port) throws GRPCException {
        DiscoveryRequest request = DiscoveryRequest.newBuilder()
                                        .setAgentId(agentId)
                                        .setAddress(address)
                                        .setPort(port)
                                        .build();
        try {
            blockingStub.discoveryAgent(request);
        } catch (StatusRuntimeException ex) {
            logger.atError().withThrowable(ex).log(GRPC_REQUEST_FAILED);
            throw new GRPCException(ex);
        }
    }

    /**
     * Unregister agent.
     *
     * @param reason reason of unregistering
     * @throws GRPCException on errors
     */
    void unregisterAgent(String reason) throws GRPCException {
        UnregisterRequest request = UnregisterRequest.newBuilder()
                                        .setAgentId(agentId)
                                        .setReason(reason)
                                        .build();
        try {
            blockingStub.unregisterAgent(request);
        } catch (StatusRuntimeException ex) {
            logger.atError().withThrowable(ex).log(GRPC_REQUEST_FAILED);
            throw new GRPCException(ex);
        }
    }


    @Override
    public void onReceiveMqttMessage(int connectionId, MqttReceivedMessage message) {
        Mqtt5Message.Builder msgBuilder = Mqtt5Message.newBuilder()
                                        .setTopic(message.getTopic())
                                        .setPayload(ByteString.copyFrom(message.getPayload()))
                                        .setQos(MqttQoS.forNumber(message.getQos()))
                                        .setRetain(message.isRetain());
        final Boolean payloadFormatIndicator = message.getPayloadFormatIndicator();
        if (payloadFormatIndicator != null) {
            msgBuilder.setPayloadFormatIndicator(payloadFormatIndicator);
        }

        final Integer messageExpiryInterval = message.getMessageExpiryInterval();
        if (messageExpiryInterval != null) {
            msgBuilder.setMessageExpiryInterval(messageExpiryInterval);
        }

        final String contentType = message.getContentType();
        if (contentType != null) {
            msgBuilder.setContentType(contentType);
        }

        final String responseTopic = message.getResponseTopic();
        if (responseTopic != null) {
            msgBuilder.setResponseTopic(responseTopic);
        }

        final byte[] correlationData = message.getCorrelationData();
        if (correlationData != null) {
            msgBuilder.setCorrelationData(ByteString.copyFrom(correlationData));
        }

        if (message.getUserProperties() != null) {
            msgBuilder.addAllProperties(message.getUserProperties());
        }

        OnReceiveMessageRequest request = OnReceiveMessageRequest.newBuilder()
                        .setAgentId(agentId)
                        .setConnectionId(MqttConnectionId.newBuilder().setConnectionId(connectionId).build())
                        .setMsg(msgBuilder.build())
                        .build();
        try {
            blockingStub.onReceiveMessage(request);
        } catch (StatusRuntimeException ex) {
            logger.atError().withThrowable(ex).log(GRPC_REQUEST_FAILED);
        }
    }

    @Override
    public void onMqttDisconnect(int connectionId, DisconnectInfo disconnectInfo, String error) {
        OnMqttDisconnectRequest.Builder builder = OnMqttDisconnectRequest.newBuilder()
                        .setAgentId(agentId)
                        .setConnectionId(MqttConnectionId.newBuilder().setConnectionId(connectionId)
                        .build());
        if (disconnectInfo != null) {
            // TODO: fill
            Mqtt5Disconnect disconnect = Mqtt5Disconnect.newBuilder().build();
            builder.setDisconnect(disconnect);
        }
        if (error != null) {
            builder.setError(error);
        }

        try {
            blockingStub.onMqttDisconnect(builder.build());
        } catch (StatusRuntimeException ex) {
            logger.atError().withThrowable(ex).log(GRPC_REQUEST_FAILED);
        }
    }

    /**
     * Closes the gRPC client.
     */
    void close() {
        channel.shutdown();
    }
}
